language/bytecode_verifier/src/type_memory_safety.rs

```diff
@@ -50,6 +50,34 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
         let function_definition_view = FunctionDefinitionView::new(module, function_definition);
         let locals_signature_view = function_definition_view.locals_signature();
         let function_signature_view = function_definition_view.signature();
+        if function_signature_view.arg_count() > locals_signature_view.len() {
+            return vec![VMStatus::new(StatusCode::RANGE_OUT_OF_BOUNDS)
+                .with_message("Fewer locals than parameters".to_string())];
+        }
+        let errors: Vec<VMStatus> = function_signature_view
```

Comment:
This is a question for the future, but: could we get rid of this logic by avoiding the duplication of the argument types? They appear in both `FunctionSignature.arg_types` and `LocalsSignature`. It seems like we could either:

- Use `FunctionSignature.arg_count` instead of arg types and require clients that want to know the arg types also have access to the `LocalsSignature`. This would let us remove the logic here, but we would still need the check above.
- Leave the arg and types in separate vectors, but view them as a single vector for the purpose of indexing. This would complicate the code that resolves local indexes a bit, but it would let us eliminate *both* this check and the check above.

To give a bit more context on the second idea: if I have 2 arguments and 5 locals, someone who asks for local index 1 would get an pointer into the arguments table, but someone who asks for local index 2 would get a pointer into the locals table.

Reply:
The second idea is closely related to the one we discussed yesterday about adding `LdArg`, `StArg`, etc., except that we wouldn't actually need to change the instruction set.

Reply:
@dariorussi is the ambassador for this issue :-)  I will let him deal with it.

Reply:
The tricky case is a function handle imported from another module. It's not defined in the current module, so no access to the locals signature. However arg types need to be known to allow the current module to be checked.

Reply:
@sblackshear yeah we will review it and all your possible solutions were mentioned above ;)
so we'll get back to this later

Reply:
@vgao1996 That is a definitely a problem with option, but not with option 2.

@dariorussi Heh, sorry I somehow missed your explanation of option 2 (which is also much more concise):
> The other is to treat them conceptually as a single array (so no changes in opcode) and just make the *Loc opcode suck (you do the math and do it right).